internal/rbd/rbd_util.go

```diff
@@ -881,6 +867,31 @@ func genVolFromVolID(ctx context.Context, volumeID string, cr *util.Credentials,
 	return rbdVol, err
 }
 
+// genVolFromVolID generates a rbdVolume structure from the provided identifier, updating
+// the structure with elements from on-disk image metadata as well.
+func genVolFromVolID(ctx context.Context, volumeID string, cr *util.Credentials, secrets map[string]string) (*rbdVolume, error) {
+	vol, err := generateVolumeFromVolumeID(ctx, volumeID, cr, secrets)
+	if errors.Is(err, util.ErrKeyNotFound) || errors.Is(err, util.ErrPoolNotFound) {
```

Comment:
Easier to do the `if` the other way around. A small body inside the statement returning `nil, err`, and continue with the rest outside the if-statement like it is the default behaviour.

When an error occurs, ideally return `nil` and an error, not some incomplete `vol` object. Make sure to call `vol.Destroy()` in error cases too.

Reply:
Need to return vol even in case of error to destroy the context and to release the lock

Reply:
really? where is the lock taken then? Ideally it should be released at the same level...